### PR TITLE
Implement EOPatch method to load and save from AWS S3 bucket

### PR DESCRIPTION
**Pull request by mlubej**
Adds EOPatch methods to save and load from AWS S3 bucket.

Still need to improve the code and unit tests, please contribute as soon as possible.

### CODE REVIEW

**Comment by mlubej** [comment]
TODO:
- [ ] Improve/extend code
- [ ] Make code more elegant 
- [ ] Make EOTasks
- [ ] Unit tests

**Comment by mlubej** [comment]
> I miss a test that I can run and also see how it was meant to be used. If a test is not there I'd have to write my own in order to properly review the code.

It's hard to make a test, since this requires some resources on the AWS bucket side to save to/load from. I don't know how to handle this.

File: core/eolearn/core/eodata.py [hunk 1]
```diff
@@ -689,6 +715,44 @@ def load(path, features=..., lazy_loading=False, mmap=False):
...
+        s3client = boto3.client('s3')
+        if not patch_location.endswith('/'):
+            patch_location += '/'
+
+        request = s3client.list_objects(Bucket=bucket_name, Prefix=patch_location)
+        fpaths = [x['Key'] for x in request['Contents']]
+
+        content = {}
+        for fpath in fpaths:
+            split = fpath[len(patch_location):].split('/')
```
[the remark this answers is not in the document]
Comment (mlubej) [reply core/eolearn/core/eodata.py]: I tried some approaches from @iovsn in the save and load functions, please check if this is more satisfactory

**Comment by mlubej** [comment]
> I miss a test that I can run and also see how it was meant to be used. If a test is not there I'd have to write my own in order to properly review the code.

Added

File: core/eolearn/core/eodata.py [hunk 2]
```diff
@@ -558,6 +559,31 @@ def save(self, path, features=..., file_format=FileFormat.NPY,
...
+        paths = ['{}/{}/{}.npy'.format(patch_location, ftype.value, fname) for ftype, fname in features]
+
+        meta = [feature for feature in self.get_features() if feature.is_meta()]
+        meta = [(ftype, None) for ftype in meta]
+        meta_paths = ['{}/{}.pkl'.format(patch_location, ftype.value) for ftype, _ in meta]
+
+        features += meta
+        paths += meta_paths
+
+        streams = (pickle.dumps(self[feat], protocol=pickle.HIGHEST_PROTOCOL) for feat in features)
```
[the remark this answers is not in the document]
Comment (mlubej) [reply core/eolearn/core/eodata.py]: On one hand it's beter to handle this with numpy, but on the other it's convenient if everything is done in one way. Do you think this is a good exception?

[the remark this answers is not in the document]
Comment (mlubej) [reply core/eolearn/core/eodata.py]: > I see you are pickle dumping data that you will upload and save under `.npy` file extension. Do you need any help figuring out how to dump numpy arrays into streams?

On one hand it's beter to handle this with numpy, but on the other it's convenient if everything is done in one way. Do you think this is a good exception?

File: core/eolearn/core/eodata.py [hunk 3]
```diff
@@ -558,6 +559,31 @@ def save(self, path, features=..., file_format=FileFormat.NPY,
...
+
+        s3client = boto3.client('s3')
+        for stream, path in zip(streams, paths):
```
[the remark this answers is not in the document]
Comment (mlubej) [reply core/eolearn/core/eodata.py]: > Credentials could be taken from `SHConfig` and if they are not set there, then they will be by default taken from aws profile. In any case I would move this into `AwsDownloadClient` in sentinelhub-py and parallelize it there.

This about `AwsDownloadClient` is a good idea, but I would do it in a separate PR. For now I added an option to pass an s3 client as an optional parameter. If it isn't provided, a default one is created from aws credentials.

Would that be satisfactory for this round?

File: core/eolearn/tests/test_aws.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
+"""
+Credits:
+Copyright (c) 2017-2019 Matej Aleksandrov, Matej Batič, Andrej Burja, Eva Erzin (Sinergise)
+Copyright (c) 2017-2019 Grega Milčinski, Matic Lubej, Devis Peresutti, Jernej Puc, Tomislav Slijepčević (Sinergise)
+Copyright (c) 2017-2019 Blaž Sovdat, Jovan Višnjić, Anže Zupanc, Lojze Žust (Sinergise)
```
Comment (mlubej) [inline core/eolearn/tests/test_aws.py]: @AleksMat is this automatically generated? If not, we should add @veseln.

File: core/eolearn/core/eodata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AleksMat) [reply core/eolearn/core/eodata.py]: In first versions of eo-learn we did use pickle for everything but it turned out using npy for numpy arrays gives you faster saving and loading and the size of saved data is smaller.

File: core/eolearn/core/eodata.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AleksMat) [reply core/eolearn/core/eodata.py]: Ok, I agree. We would have to make a few adjustments to `AwsDownloadClient` anyway. I'll make a Trello card for this.

File: core/eolearn/tests/test_aws.py (hunk 4 above)
Comment (AleksMat) [reply core/eolearn/tests/test_aws.py]: It is not automatic, we should definitely add him. Also elsewhere where he contributed.

Comment (mlubej) [reply core/eolearn/tests/test_aws.py]: Fixed in https://github.com/sentinel-hub/eo-learn/pull/165

File: core/eolearn/core/eodata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlubej) [reply core/eolearn/core/eodata.py]: The problem is not with reading/loading/compressing `numpy` arrays, but with serializing them. `numpy` uses `tobytes` and `frombuffer` to serialize arrays, but in the [documentation](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.frombuffer.html) of `frombuffer` it says that it interprets a buffer as a 1-dimensional array. Do we have EOPatch dimensions available for reshaping? Does it even make sense to do that?

**Pull request merged**